primitives/numeric/src/lib.rs

```diff
@@ -175,26 +175,41 @@ impl Numeric {
         self.inner.scale()
     }
 
-    /// Checked addition
-    ///
-    /// # Errors
-    /// In case of overflow
+    /// Checked addition. Computes `self + other`, returning `None` if overflow occurred
     pub fn checked_add(self, other: Self) -> Option<Self> {
         self.inner
             .checked_add(other.inner)
             .map(|inner| Self { inner })
     }
 
-    /// Checked subtraction
-    ///
-    /// # Errors
-    /// In case of overflow
+    /// Checked subtraction. Computes `self - other`, returning `None` if overflow occurred
     pub fn checked_sub(self, other: Self) -> Option<Self> {
         self.inner
             .checked_sub(other.inner)
             .and_then(|inner| inner.is_sign_positive().then_some(Self { inner }))
     }
 
+    /// Checked multiplication. Computes `self * other`, returning `None` if overflow occurred
+    pub fn checked_mul(self, other: Self) -> Option<Self> {
+        self.inner
+            .checked_mul(other.inner)
+            .map(|inner| Self { inner })
+    }
+
+    /// Checked division. Computes `self / other`, returning `None` if overflow occurred.
+    pub fn checked_div(self, other: Self) -> Option<Self> {
+        self.inner
+            .checked_div(other.inner)
```

Comment:
how about now?

Reply:
I think it's better to remove the validation and mention in the docs and mul and div can cause the numeric spec to change, and that users need to do their own validation as necessary. This impl assumes too much, there are multiple ways to handle it (round to `self`, round to provided `spec`, return none if accuracy goes out of bounds, etc)

What are the intended use cases of these ops?

Reply:
> What are the intended use cases of these ops?

for example, the bank wants to pay an interest to the bond holder. To do that it needs to multiply bond price with the interest rate and mint that amount for the holder. If the asset definition of the `Numeric` was defined with a certain `NumericSpec` the assignment will then fail

Reply:
Maybe instead of trying to be general with mul and rem we could provide common money-like operations on the type?

Reply:
I think that considering `NumericSpec` when doing mathematical operations is quite appropriate. Users should think about what precision their numbers are especially for financial use-cases. I also feel that having to provide explicit spec would help prevent many bugs and save development time.

I can propose a different api: `NumericSpec::checked_mul(a: Numeric, b: Numeric)` but I think the former is better

Reply:
> Maybe instead of trying to be general with mul and rem we could provide common money-like operations on the type?

we can add this on top, but there has to be something that you can depend to implement custom logic if we don't provide with the common case

Reply:
A couple of approaches we could take:

1. Expose the inner Decimal and let the user do whatever they need with it outside of Iroha's context, then validate the value when the user tries to reintroduce the value into Iroha's context.

2. Do not expose the inner Decimal and keep the Numeric always valid. This one's trickier because our idea of how operations should work might not match how a business would actually want to handle numbers. E.g. if we make calculations round, then chained operations might give users wrong results. Or maybe a user will want to use a different rounding scheme.

Reply:
> A couple of approaches we could take:
> 
>     1. Expose the inner `Decimal` and let the user do whatever they need with it outside of Iroha's context, then validate the value when the user tries to reintroduce the value into Iroha's context.

we don't want to leak internals into public api

>     2. Do not expose the inner `Decimal` and keep the `Numeric` always valid. This one's trickier because our idea of how operations should work might not match how a business would actually want to handle numbers. E.g. if we make calculations round, then chained operations might give users wrong results. Or maybe a user will want to use a different rounding scheme.

if the user gives `NumericSpec` with no scale, no rounding will happen. We can explore options in this direction

Reply:
> then chained operations might give users wrong results

It sounds reasonable if you use a fixed-precision number that the result is limited to the number of decimal points that the number can represent. If this doesn't work for you as the user you can use `NumericSpec(scale=None)` while doing calculations and then round the number afterwards

Reply:
> we don't want to leak internals into public api

It can be wrapped then into smth like `RawNumeric` with basic methods mirroring the inner `Decimal` 1-1. The point is to defer validation until the number needs to interact with Iroha.

Reply:
> The point is to defer validation until the number needs to interact with Iroha.

With the current approach you can do this by using `NumericSpec::unconstrained`. Every time you're doing an operation on 2 decimals you choose to fit the result into the spec or not

Reply:
Sure. I'll approve after you fix integration tests

Reply:
it seems that integration tests are failing regardless of my pr

Reply:
Should I open an issue for this?